src/transformers/trainer.py

```diff
@@ -358,8 +358,8 @@ def __init__(
         model: Union[PreTrainedModel, nn.Module] = None,
         args: TrainingArguments = None,
         data_collator: Optional[DataCollator] = None,
-        train_dataset: Optional[Union[Dataset, IterableDataset]] = None,
-        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
+        train_dataset: Optional[Union[Dataset, IterableDataset, "datasets.Dataset"]] = None,
```

Comment:
To do the type checking properly, I think we need at the top of the file:

```
if TYPE_CHECKING
    if is_datasets_available:
        import datasets
```

This way, the type is properly defined when the module is loaded, which isn't necessarily the case here, as datasets isn't defined